### PR TITLE
Implemented soft bodies (2)

### PR DESCRIPTION
**Pull request by yomboprime**
#125
Sorry about the repository issue.

I've been able to set the number of iterations for the simulation of the cloth and rope, and now it's much better, the jitter is gone and the soft bodies are less flexible (more like a real cloth and rope).

I did that by writing a C++ function `btSoftBody::setIterations' that sets  the iterations in the `btSoftBody::m_cfg` struct.

As I said earlier (#125), I've not been able to set the members of that struct directly, either with a reference, a pointer, or declaring as an `attribute`. So the iterations part is solved, but not in an elegant way.

Remains unsolved the question about soft-rigid collisions, I've not been able to make the rope or cloth collide with the bricks or other objects. I keep investigating.

### CODE REVIEW

**Comment by kripken** [comment]
Does the [`attribute` approach](https://github.com/kripken/emscripten/blob/master/tests/webidl/test.idl#L119) not work? You should get `get_` and `set_` methods to get and set the attribute.

**Comment by yomboprime** [comment]
I didn't realise the `get_` and `set_` methods were there. But anyways it still doesn't work :-(
I've added the struct in `btSoftBody` as an `attribute` like so:
`[Value] attribute Config m_cfg;`

And then Ive tried to set the iterations this way:
```
var sbConfig = clothSoftBody.get_m_cfg();
sbConfig.viterations = 10;
sbConfig.piterations = 10;
clothSoftBody.set_m_cfg( sbConfig );
```
But id doesn't work. If I instead call `clothSoftBody.setIterations( 10, 10 );`, then it works great.

**Comment by kripken** [comment]
`Config` is a struct? I think you need `Ref` instead of `Value`, since you want to modify it, and need a reference, not a value copy.

**Comment by yomboprime** [comment]
But I can't define it as `Ref`, when compiling it complains about that not being possible:
```
WebIDL.WebIDLError: error: Unknown extended attribute Ref on attribute, <unknown> line 745:3
  [Ref] attribute Config m_cfg;
   ^
```

**Comment by yomboprime** [comment]
Oh, also: Defining the attribute as `[Value]` works as if it was declared as `[Ref]`:
```
[Value] attribute Config m_cfg;
```
```
var sbConfig = clothSoftBody.get_m_cfg();
sbConfig.set_viterations( 10 );
sbConfig.set_piterations( 10 );
```
This works (changes the internal values of the softBody), and it is not necessary to put a `clothSoftBody.set_m_cfg( sbConfig )` at the end. Which is strange.

**Comment by yomboprime** [comment]
I've added more attributes to remove the add-hoc functions I inserted before. They work all right. But they behave as if the were `[Ref]`, when they are defined as `[Value]`. Just as before with the `m_config` attribute.

**Comment by kripken** [comment]
I see, it looks like we lack support for `Ref` on attributes? (But we do have it on methods.) That sounds like something we should fix in emscripten.

Or perhaps `Ref` and `Value` should be the same for attributes that are structs? That's possibly confusing, but you're saying that's what you see in practice? I guess since we return a wrapper around a pointer anyhow, it works either way.

File: bullet/src/BulletSoftBody/btSoftBody.h [hunk 1]
```diff
@@ -881,7 +881,25 @@ class	btSoftBody : public btCollisionObject
...
-
+    
+    // XXX AMMO local addition to bullet
+    Node* getNode(int node)
+    {
+        return &m_nodes[node];
+    }
+
+    // XXX AMMO local addition to bullet
+    Config& get_m_cfg()
```
Comment (kripken) [inline bullet/src/BulletSoftBody/btSoftBody.h]: is this still needed?

**Comment by yomboprime** [comment]
> Or perhaps Ref and Value should be the same for attributes that are structs? That's possibly confusing, but you're saying that's what you see in practice?
Yes, `[Value]` works as if it was `[Ref]`. And I can't define it as `[Ref]`, emscripten  won't let you. It says it can't convert from a pointer type to a struct type, or something alike.

All right, I'll remove those forgotten functions :-)

**Comment by kripken** [comment]
Merged, thanks!

I notice now that the cloth demo looks wrong - the cloth goes through the wall? The rope one works ok on the other hand.

Sorry for the delay, btw, I was on a trip.

**Comment by yomboprime** [comment]
Thanks!
Yeah that's the only thing that does not work: collisions between soft objects and rigid bodies.
I'll let you know if I solve it.

**Pull request merged**